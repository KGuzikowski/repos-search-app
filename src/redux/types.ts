import { ThunkAction } from 'redux-thunk'
import { Action } from 'redux'

export type AppThunk<State, ReturnType = void> = ThunkAction<ReturnType, State, unknown, Action<string>>