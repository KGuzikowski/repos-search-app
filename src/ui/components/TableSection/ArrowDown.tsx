import styled from 'styled-components'
import { primaryColor, secondaryColor } from '../../style-variables'
import { ReactComponent as ArrowIcon } from './arrow_down.svg'

const ArrowDown = styled(ArrowIcon)`
    fill: ${primaryColor};
    cursor: pointer;
    width: 40px;
    height: 40px;
    transition: fill 100ms ease-in;

    &:hover {
        fill: ${secondaryColor};
    }
`

export default ArrowDown