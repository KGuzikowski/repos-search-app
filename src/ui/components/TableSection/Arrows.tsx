import styled from 'styled-components'
import { primaryColor, secondaryColor } from '../../style-variables'
import { ReactComponent as ArrowIcon } from './arrow_down.svg'

const css = `
    fill: ${primaryColor};
    cursor: pointer;
    width: 30px;
    height: 30px;
    transition: fill 100ms ease-in;
`

export const ArrowDown = styled(ArrowIcon)`
    margin-top: -5px;
    ${css}

    &:hover {
        fill: #fff;
    }

    @media only screen and (max-width: 700px) {
        width: 30px;
        height: 30px;
    }
`

export const ArrowUp = styled(ArrowIcon)`
    transform: rotate(180deg);
    margin-bottom: -5px;
    ${css}

    &:hover {
        fill: #fff;
    }

    @media only screen and (max-width: 700px) {
        width: 30px;
        height: 30px;
    }
`

export const ArrowRight = styled(ArrowIcon)`
    ${css}
    width: 40px;
    height: 40px;
    transform: rotate(-90deg);

    &:hover {
        fill: ${secondaryColor};
    }
`

export const ArrowLeft = styled(ArrowIcon)`
    ${css}
    width: 40px;
    height: 40px;
    transform: rotate(90deg);

    &:hover {
        fill: ${secondaryColor};
    }
`

export const ArrowsDiv = styled.div`
    display: flex;
    flex-direction: column;
    // justify
`