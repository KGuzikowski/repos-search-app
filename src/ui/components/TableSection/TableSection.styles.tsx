import styled from 'styled-components'
import { primaryColor, secondaryColor, secondaryDarkerColor } from '../../style-variables'

const css = `
    padding: 10px;
    font-size: 16px;
    text-align: left;
`

export const Text = styled.p`
    font-size: 20px;

    @media only screen and (max-width: 700px) {
        font-size: 16px;
    }
`
export const TableSectionContainer = styled.div`
    width: 100%;
    display: flex;
    flex-direction: column;
    aling-items: center;
`

export const Table = styled.table`
    margin-top: 10px;
    width: 100%;
    border-collapse: collapse;
`

export const TableContainer = styled.div`
    width: 100%;
    overflow: auto;
`

export const TableHeader = styled.div`
    width: 100%;
    margin: 40px 0 20px 0;
    display: flex;
    justify-content: space-between;
    aling-items: center;

    @media only screen and (max-width: 700px) {
        flex-direction: column;
        aling-items: flex-start;
        margin: 20px 0 10px 0;
    }
`

export const Th = styled.th`
    border: 3px solid ${primaryColor};
    background: ${secondaryColor};
    ${css}
`
export const Td = styled.td<{highlighted: boolean}>`
    border: 1px solid ${primaryColor};
    ${props => props.highlighted ? `color: ${secondaryDarkerColor};` : ''}
    ${css}
`

export const ThDiv = styled.div`
    display: flex;
    justify-content: space-between;
    align-items: center;
`

export const NoRepos = styled(Text)`
    @media only screen and (max-width: 700px) {
        margin-top: 30px;
        margin-bottom: 10px;
    }
`