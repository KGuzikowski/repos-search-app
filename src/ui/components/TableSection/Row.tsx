import React from 'react'
import { repoType } from '../../../redux/repos/repo.types'
import { Td } from './TableSection.styles'

type RowProps = {
    repo: repoType,
    highlighted: boolean
}

const Row = ({ repo, highlighted }: RowProps) => (
    <tr>
        <Td highlighted={highlighted}>{repo.id}</Td>
        <Td highlighted={highlighted}>{repo.name}</Td>
        <Td highlighted={highlighted}>{repo.owner.name}</Td>
        <Td highlighted={highlighted}>{repo.stars}</Td>
        <Td highlighted={highlighted}>{repo.createdAt}</Td>
    </tr>
)

export default Row 