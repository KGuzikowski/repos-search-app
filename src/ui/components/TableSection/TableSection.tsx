import React from 'react'
import { useDispatch } from 'react-redux'
import { repoType } from '../../../redux/repos/repo.types'
import { sortReposAsc, sortReposDesc } from '../../../redux/repos/repo.actions'
import { Text, Table, TableSectionContainer, Th, ThDiv, TableHeader, TableContainer, NoRepos } from './TableSection.styles'
import Row from './Row'
import { ArrowDown, ArrowUp, ArrowsDiv } from './Arrows'
import PaginationChanger from '../Pagination/PaginationChanger'
import PageChanger from '../Pagination/PageChanger'
import { userType } from '../../../redux/user/user.types'

type TableProps = {
    user: userType | null,
    repos: repoType[] | null,
    reposNum: number,
    start: number,
    end: number
}

const TableSection = ({ user, repos, reposNum, start, end }: TableProps) => {
    const dispatch = useDispatch()
    if(repos === null || repos.length === 0)
        return <NoRepos> There are no repositories selected, use search bar above to find some. </NoRepos>

    const handleArrowDownClick = (i: number) => {
        dispatch(sortReposAsc(i))
    }

    const handleArrowUpClick = (i: number) => {
        dispatch(sortReposDesc(i))
    }

    const theadContent = ['ID', 'Repo Title', 'Owner', 'Stars', 'Created at']

    return (
        <TableSectionContainer>
            <TableHeader>
                <Text>
                    Here are repositories you requested:
                </Text>
                <PaginationChanger />
            </TableHeader>
            <TableContainer>
                <Table>
                    <thead>
                        <tr>
                            {
                                theadContent.map((col, i) => (
                                    <Th key={i}>
                                        <ThDiv>
                                            <p>{col}</p>
                                            <ArrowsDiv>
                                                <ArrowUp onClick={() => handleArrowUpClick(i)}/>
                                                <ArrowDown onClick={() => handleArrowDownClick(i)}/>
                                            </ArrowsDiv>
                                        </ThDiv>
                                    </Th>   
                                ))
                            }
                        </tr>
                    </thead>
                    <tbody>
                        {repos.map(repo => <Row key={repo.id} repo={repo} highlighted={user !== null && user.id === repo.owner.id}/>)}
                    </tbody>
                </Table>
            </TableContainer>
            <PageChanger reposNum={reposNum} start={start} end={end} />
        </TableSectionContainer>
    )
}

export default TableSection 