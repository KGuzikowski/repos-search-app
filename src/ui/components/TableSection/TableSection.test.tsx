import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import TableSection from './TableSection'
import { reposStateType } from '../../../redux/repos/repo.types'

it('renders TableSection component correctly', () => {
    const repo = {
        id: 1,
        name: 'repo',
        owner: { id: 1, name: 'Karol' },
        stars: 2,
        createdAt: '2019-05-12'
    }

    const props = {
        user: null,
        repos: [repo],
        reposNum: 1,
        start: 1,
        end: 1
    }

    const INITIAL_STATE: reposStateType = {
        repos: [repo],
        isFetching: false,
        errorMessage: '',
        name: '',
        pagination: 5,
        page: 1,
    }

    const mockStore = configureStore()
    const store = mockStore(INITIAL_STATE)

    const component = create(<Provider store={store}><TableSection {...props} /></Provider>)
    expect(component.toJSON()).toMatchSnapshot()
})