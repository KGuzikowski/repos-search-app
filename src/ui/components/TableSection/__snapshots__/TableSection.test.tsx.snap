// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TableSection component correctly 1`] = `
<div
  className="sc-AxirZ ffKkZf"
>
  <div
    className="sc-AxhUy ewixqK"
  >
    <p
      className="sc-AxjAm iRikXK"
    >
      Here are repositories you requested:
    </p>
    <div
      className="sc-fznKkj grErZq"
    >
      <p
        className="sc-fzokOt gcwYou"
      >
        Items per page:
      </p>
      <select
        className="sc-fznZeY iGswVt"
        onChange={[Function]}
      >
        <option
          value={5}
        >
          5
        </option>
        <option
          value={10}
        >
          10
        </option>
        <option
          value={15}
        >
          15
        </option>
        <option
          value={20}
        >
          20
        </option>
      </select>
    </div>
  </div>
  <div
    className="sc-AxhCb cBiJgl"
  >
    <table
      className="sc-AxiKw fQDhlB"
    >
      <thead>
        <tr>
          <th
            className="sc-AxgMl cCpECe"
          >
            <div
              className="sc-Axmtr cxWobE"
            >
              <p>
                ID
              </p>
              <div
                className="sc-fznyAO fjWxgF"
              >
                <svg
                  className="sc-fzoLsD iVbjJU"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
                <svg
                  className="sc-fzozJi kxdjke"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
              </div>
            </div>
          </th>
          <th
            className="sc-AxgMl cCpECe"
          >
            <div
              className="sc-Axmtr cxWobE"
            >
              <p>
                Repo Title
              </p>
              <div
                className="sc-fznyAO fjWxgF"
              >
                <svg
                  className="sc-fzoLsD iVbjJU"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
                <svg
                  className="sc-fzozJi kxdjke"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
              </div>
            </div>
          </th>
          <th
            className="sc-AxgMl cCpECe"
          >
            <div
              className="sc-Axmtr cxWobE"
            >
              <p>
                Owner
              </p>
              <div
                className="sc-fznyAO fjWxgF"
              >
                <svg
                  className="sc-fzoLsD iVbjJU"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
                <svg
                  className="sc-fzozJi kxdjke"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
              </div>
            </div>
          </th>
          <th
            className="sc-AxgMl cCpECe"
          >
            <div
              className="sc-Axmtr cxWobE"
            >
              <p>
                Stars
              </p>
              <div
                className="sc-fznyAO fjWxgF"
              >
                <svg
                  className="sc-fzoLsD iVbjJU"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
                <svg
                  className="sc-fzozJi kxdjke"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
              </div>
            </div>
          </th>
          <th
            className="sc-AxgMl cCpECe"
          >
            <div
              className="sc-Axmtr cxWobE"
            >
              <p>
                Created at
              </p>
              <div
                className="sc-fznyAO fjWxgF"
              >
                <svg
                  className="sc-fzoLsD iVbjJU"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
                <svg
                  className="sc-fzozJi kxdjke"
                  onClick={[Function]}
                >
                  arrow_down.svg
                </svg>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="sc-AxheI gNnDlL"
          >
            1
          </td>
          <td
            className="sc-AxheI gNnDlL"
          >
            repo
          </td>
          <td
            className="sc-AxheI gNnDlL"
          >
            Karol
          </td>
          <td
            className="sc-AxheI gNnDlL"
          >
            2
          </td>
          <td
            className="sc-AxheI gNnDlL"
          >
            2019-05-12
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="sc-fzqBZW ewoBuA"
  >
    <div>
      2
      -
      1
       of 
      1
       repos
    </div>
    <div
      className="sc-fzoXzr iBuFyX"
    >
      <svg
        className="sc-fzplWN isUuJy"
        onClick={[Function]}
      >
        arrow_down.svg
      </svg>
      <p
        className="sc-fzqNJr ixBbhF"
        onClick={[Function]}
      >
        1
      </p>
      <p
        className="sc-fzoLag cAtFyR"
      >
        of
      </p>
      <p
        className="sc-fzqNJr ixBbhF"
        onClick={[Function]}
      >
        NaN
      </p>
      <svg
        className="sc-fzpans jfCHoO"
        onClick={[Function]}
      >
        arrow_down.svg
      </svg>
    </div>
  </div>
</div>
`;
